client/rest-high-level/src/main/java/org/elasticsearch/client/security/user/privileges/ClusterPrivilege.java

```diff
@@ -0,0 +1,92 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.client.security.user.privileges;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.regex.Pattern;
+
+public final class ClusterPrivilege {
+
+    private static final Pattern ALL_CLUSTER_PATTERN = Pattern.compile("^cluster:|indices:admin/template/");
+    private static final Map<String, ClusterPrivilege> builtins = new HashMap<>();
+
+    public static final ClusterPrivilege NONE = new ClusterPrivilege("none", true);
+    public static final ClusterPrivilege ALL = new ClusterPrivilege("all", true);
+    public static final ClusterPrivilege MONITOR = new ClusterPrivilege("monitor", true);
+    public static final ClusterPrivilege MONITOR_ML = new ClusterPrivilege("monitor_ml", true);
+    public static final ClusterPrivilege MONITOR_WATCHER = new ClusterPrivilege("monitor_watcher", true);
+    public static final ClusterPrivilege MONITOR_ROLLUP = new ClusterPrivilege("monitor_rollup", true);
+    public static final ClusterPrivilege MANAGE = new ClusterPrivilege("manage", true);
+    public static final ClusterPrivilege MANAGE_ML = new ClusterPrivilege("manage_ml", true);
+    public static final ClusterPrivilege MANAGE_WATCHER = new ClusterPrivilege("manage_watcher", true);
+    public static final ClusterPrivilege MANAGE_ROLLUP = new ClusterPrivilege("manage_rollup", true);
+    public static final ClusterPrivilege MANAGE_IDX_TEMPLATES = new ClusterPrivilege("manage_index_templates", true);
+    public static final ClusterPrivilege MANAGE_INGEST_PIPELINES = new ClusterPrivilege("manage_ingest_pipelines", true);
+    public static final ClusterPrivilege TRANSPORT_CLIENT = new ClusterPrivilege("transport_client", true);
+    public static final ClusterPrivilege MANAGE_SECURITY = new ClusterPrivilege("manage_security", true);
+    public static final ClusterPrivilege MANAGE_SAML = new ClusterPrivilege("manage_saml", true);
+    public static final ClusterPrivilege MANAGE_PIPELINE = new ClusterPrivilege("manage_pipeline", true);
+    public static final ClusterPrivilege MANAGE_CCR = new ClusterPrivilege("manage_ccr", true);
+    public static final ClusterPrivilege READ_CCR = new ClusterPrivilege("read_ccr", true);
+
+    private final String name;
+
+    private ClusterPrivilege(String name, boolean builtin) {
+        this.name = name;
+        if (builtin) {
+            builtins.put(name, this);
+        }
+    }
+
+    public static ClusterPrivilege fromString(String name) {
+        Objects.requireNonNull(name);
+        final ClusterPrivilege builtin = builtins.get(name);
+        if (builtin != null) {
+            return builtin;
+        }
+        if (false == ALL_CLUSTER_PATTERN.matcher(name).matches()) {
+            throw new IllegalArgumentException("[" + name + "] is not a cluster action privilege.");
```

Comment:
Interesting 🤔 

> It means if we add a new cluster privilege then it needs to be added to the HLRC

There is no need to update the client in this case, unless you need this new privilege that we added on the server. If you need this, then you have to change client code as well, and since you're there why not update the client library as well? (aside that COBOL developers might prefer the low level client instead)

> From a separation of concerns point of view, why should the client know/care which strings are understood by the server, and duplicate that logic?

I believe it kinda should do. This is what *high-level* client means in my head. It enforces some types, but *sometimes* it enforces some values too.
Tying on a previous discussion, why should we validate for *null* values on the client side, at all? Because having a null *username* or *password* for example, would be invalidated on the server.

I think this particular example is vulnerable to this type of argument, because the type and value mingle. What if instead of the formatted strings, we would've had a naming scheme based on `Enums`? Would we be duplicating the `Enums` on the client or would we be using Strings?

> The worst case for this will be if an older client cannot even retrieve the list of roles from the cluster because one of the roles contains a privilege string that the client rejects.

I agree, this is the worst case. But even if the client would be able to deserialize they would not be useful, because the client cannot possibly understand them. There is no solution to infinite forward compatibility in the high level client. You can always stick with the low level client.

The alternative we all have in our heads is using Strings. It's the alternative low level client uses.

Reply:
If this is something that should be enforced like this, we should turn it into an enum instead.

Reply:
I have made `ClusterPrivilege` and `IndexPrivilege` strings. The previous classes did not hold any other information besides the string value and there was no reason to validate the format of the privilege value on the client.

Being able to construct a privilege on the client side is a higher clients business.

Reply:
For what it's worth, I think there's friendliness in defining constants for the Strings for that developers can use `IndexPrivilege.READ`, but I don't think the client needs to enforce that you only use the enumerated values.

Especially since the server doesn't even care about that...
```
PUT _xpack/security/role/test1
{ "cluster": [ "foobar" ] }

{"role":{"created":true}}

GET _xpack/security/role/test1

{"test1":{
  "cluster":["foobar"],
  "indices":[],
  "applications":[],
  "run_as":[],
  "metadata":{},
  "transient_metadata":{"enabled":true}
}}
```

Reply:
wow, this is news to me. It will throw when trying to build the role out of the role descriptors (most def - looked at the code, but did not test it), but this is unexpected... TIL
Thank you for pointing this out, to me!